src/work/WorkSequence.h

```diff
@@ -17,13 +19,16 @@ namespace stellar
 class WorkSequence : public BasicWork
 {
     std::vector<std::shared_ptr<BasicWork>> mSequenceOfWork;
+    std::vector<ConditionFn> mConditions;
     std::vector<std::shared_ptr<BasicWork>>::const_iterator mNextInSequence;
     std::shared_ptr<BasicWork> mCurrentExecuting;
 
   public:
-    WorkSequence(Application& app, std::string name,
-                 std::vector<std::shared_ptr<BasicWork>> sequence,
-                 size_t maxRetries = RETRY_A_FEW);
+    WorkSequence(
+        Application& app, std::string name,
+        std::vector<std::shared_ptr<BasicWork>> sequence,
```

Comment:
it looks like you expect each sequence to have an optional precondition, the contract should instead take a vector of pairs and a comment should explain what they are for

Reply:
actually there is a better way: you should just create a `GatedWork` (that derives from `BasicWork`) that takes a condition and `BasicWork` as a parameter, and when you need a specific work to be gated by some arbitrary condition, you just "wrap it" using that `GatedWork`.

It waits for the condition to be met (sleeps for 1 second if not), and when met, it "unblocks" itself and invokes the `BasicWork`.

Reply:
(notes from online and offline discussions)
I think the GatedWork approach is cleaner, my only concern is that we might have a lot of work waiting and waking up, especially for checkpoints that take a while to apply. This is not the case for current download-apply work, which wakes up its waiting children. What I don't like about the "parent wakes its child" approach is that the child can't recover (succeed or fail) on its own - it has to be woken up by its parent.